source/common/common/logger.cc

```diff
@@ -158,16 +158,20 @@ void DelegatingLogSink::setTlsDelegate(SinkDelegate* sink) { *tlsSink() = sink;
 SinkDelegate* DelegatingLogSink::tlsDelegate() { return *tlsSink(); }
 
 static Context* current_context = nullptr;
+static absl::Mutex current_context_mutex;
 
 Context::Context(spdlog::level::level_enum log_level, const std::string& log_format,
                  Thread::BasicLockable& lock, bool should_escape, bool enable_fine_grain_logging)
     : log_level_(log_level), log_format_(log_format), lock_(lock), should_escape_(should_escape),
-      enable_fine_grain_logging_(enable_fine_grain_logging), save_context_(current_context) {
+      enable_fine_grain_logging_(enable_fine_grain_logging) {
+  absl::MutexLock context_lock(&current_context_mutex);
+  save_context_ = current_context;
   current_context = this;
   activate();
```

Comment:
Fair point. We need lines 168 and 169 (saving/setting current_context) to be under the lock but maybe not activate. That being said, it seems safer to do it that way, lest we end up with some inconsistent state.

Reply:
getFineGrainLogContext is actually a function in another file which doesn't take this lock.  I guess I tend to agree with ryan, if I read his comment correctly, that it seems safer to hold the lock while running `activate()`. 

This also keeps consistency with the other `activate()` call in the deletion, which seems like it requires the lock.

Reply:
Right, but locks in C++ are not re-entrant.  So when calling `activate()`, we are still holding onto the lock guarded by `current_context_mutex`, and then in `activate()`, we call `getFineGrainLogContext()` which tries to acquire the lock guarded by the same mutex again.  Since it's not re-entrant, won't we deadlock?

Reply:
Chatted offline, getFineGrainLogContext doesn't do what it sounds like and is ok

Reply:
Sorry it was my bad, I misread `getFineGrainLogContext()` as `getFineGrainLogFormat()` and thought it was in the same logger.cc file guarded by the same mutex, but that's not the case, so the question of re-entrant locks doesn't arise.